File: android_app/app/src/main/java/com/oceansentinels/app/domain/repository/IncidentRepository.kt
package com.oceansentinels.app.domain.repository

import com.oceansentinels.app.domain.model.*
import kotlinx.coroutines.flow.Flow

/**
 * Repository interface for incident operations
 */
interface IncidentRepository {
    
    /**
     * Get all incidents with optional filters
     */
    suspend fun getIncidents(filters: IncidentFilters): Result<IncidentListResult>
    
    /**
     * Get incidents as a flow for real-time updates
     */
    fun getIncidentsFlow(filters: IncidentFilters): Flow<Result<IncidentListResult>>
    
    /**
     * Get a single incident by ID
     */
    suspend fun getIncident(id: Int): Result<Incident>
    
    /**
     * Get a single incident by reference ID
     */
    suspend fun getIncidentByReference(referenceId: String): Result<Incident>
    
    /**
     * Get user's own reports
     */
    suspend fun getMyReports(filters: IncidentFilters): Result<IncidentListResult>
    
    /**
     * Create a new incident report
     */
    suspend fun createIncident(request: CreateIncidentRequest): Result<Incident>
    
    /**
     * Verify an incident (Admin/Authority only)
     */
    suspend fun verifyIncident(id: Int): Result<Incident>
    
    /**
     * Deploy response to an incident (Admin/Rescue Team only)
     */
    suspend fun deployResponse(id: Int): Result<Incident>
    
    /**
     * Resolve an incident (Admin/Authority/Rescue Team only)
     */
    suspend fun resolveIncident(id: Int): Result<Incident>
    
    /**
     * Get cached incidents from local database
     */
    suspend fun getCachedIncidents(): List<Incident>
    
    /**
     * Sync local incidents with server
     */
    suspend fun syncIncidents(): Result<Unit>
    
    /**
     * Assign an incident to a rescue team member (Admin/Authority only)
     */
    suspend fun assignIncident(incidentId: Int, rescueTeamUserId: Int): Result<Unit>
    
    /**
     * Get incidents assigned to the current rescue team member
     */
    suspend fun getAssignedIncidents(page: Int = 1, size: Int = 20): Result<IncidentListResult>
}

File: IOS/OceanSentinels/Domain/Repositories/IncidentRepository.swift
import Foundation

/// Repository protocol for incident management operations.
protocol IncidentRepository {

    /// Fetches a paginated, filtered list of incidents.
    func getIncidents(filters: IncidentFilters) async throws -> IncidentListResult

    /// Fetches a single incident by its ID.
    func getIncident(id: Int) async throws -> Incident

    /// Fetches an incident by its reference ID from cache.
    func getIncidentByReference(referenceId: String) async throws -> Incident

    /// Fetches the current user's own reports.
    func getMyReports(filters: IncidentFilters) async throws -> IncidentListResult

    /// Creates a new incident report.
    func createIncident(request: CreateIncidentRequest) async throws -> Incident

    /// Marks an incident as verified. Returns the updated incident.
    func verifyIncident(id: Int) async throws -> Incident

    /// Deploys a response to a verified incident. Returns the updated incident.
    func deployResponse(id: Int) async throws -> Incident

    /// Marks an in-progress incident as resolved. Returns the updated incident.
    func resolveIncident(id: Int) async throws -> Incident

    /// Assigns an incident to a rescue team member.
    func assignIncident(incidentId: Int, rescueTeamUserId: Int) async throws

    /// Fetches incidents assigned to the current user.
    func getAssignedIncidents(page: Int, size: Int) async throws -> IncidentListResult

    /// Returns locally cached incidents.
    func getCachedIncidents() async throws -> [Incident]

    /// Syncs local incidents with the server.
    func syncIncidents() async throws

    /// Uploads a photo for an incident.
    func uploadPhoto(imageData: Data, fileName: String) async throws -> String
}
